### ISSUE TITLE
Correctly implement function equality testing within structures

### ISSUE THREAD

**Issue by jwiegley** [haskell-nix/hnix#340]
This issue tracks the fact that the following tests in `EvalTests.hs` fail when `MATCHING_TESTS=yes`:

  - [ ] `case_function_equals1`
  - [ ] `case_function_equals2`

**Comment by matthewbauer** [comment]
Hmmm... I'm confused by the error:

```
    function equals2:
FAIL (0.02s)
      tests/TestCommon.hs:65:
      ([(x: x)] == [(x: x)])
      expected: "false\n"
       but got: "true\n"
```

But  ```([(x: x)] == [(x: x)])``` should be true and the test case is ```constantEqualText "true" "([(x: x)] == [(y: y)])"```. Where does the false come from?

**Comment by ryantrinkle** [comment]
I think it comes from the the C++ implementation:

```
» nix-repl
Welcome to Nix version 1.11.16. Type :? for help.

nix-repl> ([(x: x)] == [(x: x)])
false
```

**Comment by matthewbauer** [comment]
Very weird. Nix has very weird behaviors. I will fix it with the test mentioned in #323.

**Comment by matthewbauer** [comment]
Hnix isn't running Nix in those tests. I think I was just thinking about it backwards the left side is the actual, the right is the expected.
